libafl_intelpt/src/lib.rs

```diff
@@ -0,0 +1,1026 @@
+    }
+    #[cfg(not(target_os = "linux"))]
+    reasons.push("Only linux hosts are supported at the moment".to_owned());
+
+    if reasons.is_empty() {
+        Ok(())
+    } else {
+        Err(reasons.join("; "))
+    }
+}
+
+/// Convert [`PtError`] into [`Error`]
+#[inline]
+#[must_use]
+pub fn error_from_pt_error(err: PtError) -> Error {
+    Error::unknown(err.to_string())
+}
+
+#[cfg(target_os = "linux")]
+fn availability_in_linux() -> Result<(), String> {
+    let mut reasons = Vec::new();
+    match linux_version() {
+        // https://docs.rs/perf-event-open-sys/4.0.0/perf_event_open_sys/#kernel-versions
+        Ok(ver) if ver >= (5, 19, 4) => {}
+        Ok((major, minor, patch)) => reasons.push(format!(
+            "Kernel version {major}.{minor}.{patch} is older than 5.19.4 and might not work."
+        )),
+        Err(()) => reasons.push("Failed to retrieve kernel version".to_owned()),
+    }
+
+    if let Err(e) = &*PERF_EVENT_TYPE {
+        reasons.push(e.clone());
+    }
+
+    if let Err(e) = &*NR_ADDR_FILTERS {
+        reasons.push(e.clone());
+    }
+
+    // official way of knowing if perf_event_open() support is enabled
+    // https://man7.org/linux/man-pages/man2/perf_event_open.2.html
+    let perf_event_support_path = "/proc/sys/kernel/perf_event_paranoid";
+    if !Path::new(perf_event_support_path).exists() {
+        reasons.push(format!(
+            "perf_event_open() support is not enabled: {perf_event_support_path} not found"
+        ));
+    }
+
+    // TODO check also the value of perf_event_paranoid, check which values are required by pt
+    // https://www.kernel.org/doc/Documentation/sysctl/kernel.txt
+    // also, looks like it is distribution dependent
+    // https://askubuntu.com/questions/1400874/what-does-perf-paranoia-level-four-do
+    // CAP_SYS_ADMIN might make this check useless
+
+    match caps::read(None, CapSet::Permitted) {
+        Ok(current_capabilities) => {
+            let required_caps = [
+                Capability::CAP_IPC_LOCK,
+                Capability::CAP_SYS_PTRACE,
+                Capability::CAP_SYS_ADMIN, // TODO: CAP_PERFMON doesn't look to be enough!?
+                Capability::CAP_SYSLOG,
+            ];
+
+            for rc in required_caps {
+                if !current_capabilities.contains(&rc) {
+                    reasons.push(format!("Required capability {rc} missing"));
+                }
+            }
+        }
+        Err(e) => reasons.push(format!("Failed to read linux capabilities: {e}")),
+    };
+
+    if reasons.is_empty() {
+        Ok(())
+    } else {
+        Err(reasons.join("; "))
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn new_perf_event_attr_intel_pt() -> Result<perf_event_attr, Error> {
+    let type_ = match &*PERF_EVENT_TYPE {
+        Ok(t) => Ok(*t),
+        Err(e) => Err(Error::unsupported(e.clone())),
+    }?;
+    let config = PtConfig::builder()
+        .with_noretcomp(true)
+        .with_psb_period(u4::new(0))
```

Comment:
Also here there's the question if this can't be done with that bitfield trait instead of grabbing yet another dep? Just asking out loud :)

Reply:
I'm not sure I've understood :D
If you mean: _do we really need `arbitrary-int`_?
Then yes, the bitbybit crate is based on it and has it as a dependency anyway...

Reply:
Yeah that was the question. Thanks

So the question remains:: what's better about bitbybit? And can you change the other bit fields to it if you really want it (I don't care either way but don't want multiple crates for the same thing)

Reply:
I see, for the moment I changed the code using it and commented out the dependency.
It was used to just build a constant so it's not a big deal. I will introduce it in the future with a new PR changing also the `c2rust-bitfields`.
(In the future the `PtConfig` bitfield struct will potentially support more options)